### PR TITLE
coll: automatic enums for collective algorithm cvars

### PR DESCRIPTION
**Pull request by hzhou**
Separate PR for the cvar refactor from #3603.

### CODE REVIEW

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by hzhou** [comment]
test:jenkins/ch3/mxm

**Comment by hzhou** [comment]
test:jenkins/ch3/mxm

**Comment by hzhou** [comment]
test:jenkins/ch4/ucx

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by hzhou** [comment]
Had to disable the enum error check as there is no failure mode defined in the `MPIR_T_..._int` interface right now.

test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by wesbland** [comment]
I'm still reviewing the contents of the patch, but just looking at the docs patch, I'd say that a description of the format that the CVAR description is still needed (beyond just saying that it's YAML).

**Comment by hzhou** [comment]
> I'm still reviewing the contents of the PR, but just looking at the docs patch, I'd say that a description of the format that the CVAR description is still needed (beyond just saying that it's YAML).

Do you have specific suggestion or questions that I can add? I guess I can add emphasis that indentation is important and multi-line string format in description has a special syntax. The cvar enum hack depends on listing all the enum items in description in the format e.g.:
    entry1    - ...
    entry2    - ...
, which I definitely need document, as well as comment in `extractcvars.in`.

**Comment by wesbland** [comment]
> Do you have specific suggestion or questions that I can add? I guess I can add emphasis that indentation is important and multi-line string format in description has a special syntax. The cvar enum hack depends on listing all the enum items in description in the format e.g.:
> 
> ```
>     entry1    - ...
>     entry2    - ...
> ```
> , which I definitely need document, as well as comment in `extractcvars.in`.

The last bit in particular is what I was looking for. Also, are there any particular rules about the lines of documentation that _aren't_ part of the enum values (e.g. the number of lines or what they start with)?

**Comment by hzhou** [comment]
> Also, are there any particular rules about the lines of documentation that _aren't_ part of the enum values (e.g. the number of lines or what they start with)?

Only the pattern /^\s*(\w+)\s*-\s*/ inside description is recognized as enum values, anything else can freely go anywhere as long as it won't misfit that regex.

**Comment by wesbland** [comment]
> Only the pattern /^\s*(\w+)\s*-\s*/ inside description is recognized as enum values, anything else can freely go anywhere as long as it won't misfit that regex.

That's good. 👍 Maybe just a simple line saying that in the docs would be sufficient.

File: src/mpi/coll/barrier/barrier.c [hunk 1]
```diff
@@ -22,20 +22,20 @@
...
       verbosity   : MPI_T_VERBOSITY_USER_BASIC
       scope       : MPI_T_SCOPE_ALL_EQ
       description : |-
         Variable to select barrier algorithm
-        auto               - Internal algorithm selection
-        nb                 - Force nonblocking algorithm
-        recursive_doubling - Force recursive doubling algorithm
+        auto          - Internal algorithm selection
+        nb            - Force nonblocking algorithm
+        dissemination - Force recursive doubling algorithm
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpi/coll/barrier/barrier.c]: I agree.

File: src/mpi/coll/ireduce/ireduce.c [hunk 2]
```diff
@@ -82,12 +82,12 @@
         auto                  - Internal algorithm selection
         binomial              - Force binomial algorithm
         reduce_scatter_gather - Force reduce scatter gather algorithm
-        tree                  - Force Generic Transport Tree
-        ring                  - Force Generic Transport Ring
+        gentran_tree          - Force Generic Transport Tree
+        gentran_ring          - Force Generic Transport Ring
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpi/coll/ireduce/ireduce.c]: LOL Now you are sharing my pain :)

File: src/mpi/coll/iscatter/iscatter.c [hunk 3]
```diff
@@ -12,16 +12,16 @@
...
       verbosity   : MPI_T_VERBOSITY_USER_BASIC
       scope       : MPI_T_SCOPE_ALL_EQ
       description : |-
         Variable to select iscatter algorithm
-        auto     - Internal algorithm selection
-        binomial - Force binomial algorithm
-        tree     - Force genetric transport based tree algorithm
+        auto         - Internal algorithm selection
+        binomial     - Force binomial algorithm
+        gentran_tree - Force genetric transport based tree algorithm
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpi/coll/iscatter/iscatter.c]: Appreciate you marking all them out for me!

**Comment by hzhou** [comment]
test:jenkins/ch3/mxm

**Comment by hzhou** [comment]
@wesbland I have split the commit. However, the cvar enum patch necessitates the addition of missing descriptions -- because it affects the code now, and it also necessitates the selection code change, as it introduces new enum names. While renaming the enum values into gentran prefix is in nature a separate change, but really separating them involves making changes to the same code places back and forth. The current split makes reading each commit clean, but functionally they still need be taken together. To really split them by function will involve quite some effort and will make the diffs really messy. Let me know if you insist.

**Comment by zhenggb72** [comment]
Assuming we will add error checking later, I am ready to approve it.

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by hzhou** [comment]
@wesbland upon completion of tests, can I add your sign-offs?

**Comment by raffenet** [comment]
This looks OK to me. But can we squash 71c0e5d86ffb3e73483b81c68194f201777295a2, 7641b363b844f881dcd181ce01326b77ad0e46a5, and 16e58e8711445c37e224bf3634875352ff3310ea into a single commit? It doesn't look to me like they can work individually.

**Comment by raffenet** [comment]
> This looks OK to me. But can we squash [71c0e5d](https://github.com/pmodels/mpich/commit/71c0e5d86ffb3e73483b81c68194f201777295a2), [7641b36](https://github.com/pmodels/mpich/commit/7641b363b844f881dcd181ce01326b77ad0e46a5), and [16e58e8](https://github.com/pmodels/mpich/commit/16e58e8711445c37e224bf3634875352ff3310ea) into a single commit? It doesn't look to me like they can work individually.

@wesbland do you have any objection to squashing these? It's a little gross that we essentially break the misnamed gentran functions and then fix them in the next commit, but squashing these doesn't change that, and I'd rather not spend any more cycles trying to re-order these.

**Comment by wesbland** [comment]
That’s fine.

**Pull request merged**